File: src/main/scala/my/spark/examples/scala/AverageWordLength.scala
package my.spark.examples.scala

import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf

object AverageWordLength {
  def main(args: Array[String]) {
    if (args.length != 1) {
      println("Usage: AverageWordLength <file or directory>")
      System.exit(1)
    }

    val sc = new SparkContext()
    val totalWords = sc.accumulator(0)
    val totalLetters = sc.accumulator(0.0)
    val words = sc.textFile(args(0)).flatMap(_.split(" "))
    words.foreach(word => {
       totalWords += 1
       totalLetters += word.length
    })
    println( "Average word length: " + totalLetters.value/totalWords.value)
    sc.stop()
  }
}

File: src/main/python/AverageWordLength.py
import sys
from pyspark import SparkContext

if __name__ == "__main__":
  if len(sys.argv) != 2:
    print >> sys.stderr, "Usage: AverageWordLengthCalculator <file or directory>"
    exit(-1)
  
  sc = SparkContext()
  def addTotals(word,words,letters):    
    words +=1
    letters += len(word)
  totalWords = sc.accumulator(0)
  totalLetters = sc.accumulator(0.0)
  words = sc.textFile(sys.argv[1]).flatMap(lambda line: line.split())
  words.foreach(lambda word: addTotals(word,totalWords,totalLetters))
  print "Average word length: ", totalLetters.value/totalWords.value
